File: test-suite-kotlin/src/test/kotlin/io/micronaut/mqtt/docs/properties/PropertiesSpec.kt
package io.micronaut.mqtt.docs.properties

import io.micronaut.mqtt.AbstractMQTTTest
import org.awaitility.Awaitility.await
import org.junit.jupiter.api.Test
import java.util.concurrent.TimeUnit

class PropertiesSpec : AbstractMQTTTest() {

    @Test
    fun testPublishingAndReceivingProperties() {
        val applicationContext = startContext()

        // tag::producer[]
        val productClient = applicationContext.getBean(ProductClient::class.java)
        productClient.send("body".toByteArray())
        productClient.send("guest", "text/html", "body2".toByteArray())
        productClient.send("guest", null, "body3".toByteArray())
        // end::producer[]

        val productListener = applicationContext.getBean(ProductListener::class.java)

        try {
            await().atMost(5, TimeUnit.SECONDS).until {
                productListener.messageProperties.size == 3 &&
                        productListener.messageProperties.contains("guest|application/json|myApp") &&
                        productListener.messageProperties.contains("guest|text/html|myApp") &&
                        productListener.messageProperties.contains("guest|null|myApp")
            }
        } finally {
            applicationContext.close()
        }
    }
}

File: test-suite-python/src/test/python/micronaut/mqtt/docs/properties/PropertiesSpec.py
from time import sleep
from typing import Annotated

from jakarta.inject import Inject
from micronaut.context.annotation import Property
from micronaut.test.extensions.junit5.annotation import MicronautTest
from org.junit.jupiter.api import Test

from .ProductClient import ProductClient
from .ProductListener import ProductListener


@Property(name="spec.name", value="PropertiesSpec")
@MicronautTest(environments=["mqtt"])
class PropertiesSpec:
    product_client: Annotated[ProductClient, Inject]
    product_listener: Annotated[ProductListener, Inject]

    @Test
    def test_publishing_and_receiving_properties(self):
        # tag::producer[]
        self.product_client.send(b"body")
        self.product_client.send_with_properties("guest", "text/html", b"body2")
        self.product_client.send_with_properties("guest", None, b"body3")
        # end::producer[]

        listener = self.product_listener
        for _ in range(50):
            if len(listener.message_properties) == 3:
                break
            sleep(0.1)
        assert len(listener.message_properties) == 3
        assert "guest|application/json|myApp" in listener.message_properties
        assert "guest|text/html|myApp" in listener.message_properties
        assert "guest|None|myApp" in listener.message_properties
